### PR TITLE
feat(windows): Use Keyboard Activated event API call to turn off caps lock (if required) 🚏

### PR DESCRIPTION
**Pull request by rc-swag**
Add external event processing to the kmx processor. 

At this point, the current modifier state has not been passed with the event. It could be passed as the third object with the void* pointer. Currently, the kmx_processor will queue a action to turn caps lock off, if the just loaded keyboard has the  `CapsAlwaysOff` store set, regardless of the current state of the capslock on the platform. 

Fixes: #5822 

# User Testing

For these tests, use a keyboard with the [`caps_always_off.kmp.zip`](https://github.com/keymanapp/keyman/files/8937257/caps_always_off.kmp.zip) store set. We call this keyboard _capsalwaysoff_ below.

Any keyboard with that store set will work; if you don't have one at hand you can use the `caps_always_off.kmp` keyboard. The *caps_always_off.kmp* keyboard will prevent switching caps lock on. As a sanity check to verify that Keyman is actually active, pressing the key `a` will output `ncaps_little_a`, and `Shift+a` will output `ncaps_shift_A`.

**Note:** When testing in a virtual machine, use an on-screen keyboard (in VirtualBox: Input/Keyboard/Soft Keyboard) and observe the caps lock indicator of the on-screen keyboard. Using the hardware keyboard might show side effects with caps lock.

- Install a keyboard that has `CapsAlwaysOff` store set, e.g. `caps_always_off.kmp`.
- Currently active keyboard is a non-Keyman keyboard

- **TEST_CAPSOFF-EXT_EVENT**: switching turns off caps lock
  - turn on caps lock
  - switch to capsalwaysoff keyboard
  Expected result:
  - caps lock indicator turned off

## The following tests are more regression tests to sure existing behaviour still works. 

- **TEST_CAPSOFF-1**: sanity check
  - switch to capsalwaysoff keyboard
  - press and release `a`

  Expected result:
  - output: `ncaps_little_a`

- **TEST_CAPSOFF-2**: caps lock stays off
  - switch to capsalwaysoff keyboard
  - press and release `CapsLock` key
  - press and release `a`

  Expected result:
  - caps lock indicator stays turned off
  - output: `ncaps_little_a`

- **TEST_CAPSOFF-3**: no caps lock while holding capslock key
  - switch to capsalwaysoff keyboard
  - press and hold `CapsLock` key
  - press and release `a`
  - release `CapsLock` key

  Expected result:
  - output: `ncaps_little_a`

- **TEST_CAPSOFF-4**: no caps lock while holding capslock key
  - switch to capsalwaysoff keyboard
  - press and hold `CapsLock` key
  - press and hold `Shift` key
  - press and release `a`
  - release `CapsLock` and `Shift` keys

  Expected result:
  - output: `ncaps_shift_A`

## SHIFT: CapsOnOnly/ShiftFreesCaps

For these tests, use a keyboard with the `CapsOnOnly` and `ShiftFreesCaps` stores set. We call this keyboard _shift_frees_caps_ below.

Any keyboard with these stores set will work; if you don't have one at hand you can use the [`shift_frees_caps.kmp.zip`](https://github.com/keymanapp/keyman/files/8937320/shift_frees_caps.kmp.zip) keyboard.

The _shift_frees_caps.kmp_ keyboard will enable caps lock by pressing the `CapsLock` key, and will turn capslock off by pressing the `Shift` key. The keyboard outputs _pass_ or _fail_ if following the test cases.

**Note:** When testing in a virtual machine, use an on-screen keyboard (in VirtualBox: Input/Keyboard/Soft Keyboard) and observe the caps lock indicator of the on-screen keyboard. Using the hardware keyboard might show side effects with caps lock. **Except** for TEST_CAPSONLY-5 which can only be reliably tested on a hardware keyboard on host OS (not a VM). For windows 10 and windows 11 with a virtual box vm-onscreen keyboard, the following happens. The VM soft keyboard does NOT actually send the Shift Shift Key Stroke through but rather will change the keys pressed for example if an `a` is pressed the soft keyboard itself will change that key to a `A`. This means we can't Test TEST_CAPONLY-5 on a soft keyboard. 

### Prerequisites before each test

- Install a keyboard that has the `CapsOnOnly` and `ShiftFreesCaps` stores set, e.g.
   `shift_frees_caps.kmp`.
- CapsLock is currently off
- Currently active keyboard is _shift_frees_caps_ keyboard

### Test cases

<details>
<summary>click to expand</summary>

- **TEST_CAPSONLY-1**: no caps
  - press and release `1`

  Expected result:
  - output: `pass.`

- **TEST_CAPSONLY-2**: caps
  - press and release `CapsLock`
  - press and release `2`

  Expected result:
  - caps lock indicator turned on
  - output: `pass.`

- **TEST_CAPSONLY-3**: caps doesn't toggle
  - press and release `CapsLock`
  - press and release `CapsLock`
  - press and release `6`

  Expected result:
  - caps lock indicator turned on
  - output: `pass.`

- **TEST_CAPSONLY-4**: shift turns off
  - press and release `CapsLock`
  - press and hold `Shift`
  - press and release `3`
  - release `Shift`

  Expected result:
  - caps lock indicator turned off
  - output: `pass.`

- **TEST_CAPSONLY-5**: shift by itself turns off
  Be aware of limitations when testing this on virtual machines as noted above. 
  - press and release `CapsLock`
  - press and release `Shift`

  Expected result:
  - caps lock indicator turned off
  - (no output)

</details>

### CODE REVIEW

File: core/src/kmx/kmx_processevent.h [hunk 1]
```diff
@@ -107,6 +106,8 @@ class KMX_ProcessEvent {
   KMX_Environment *GetEnvironment();
   KMX_Environment const *GetEnvironment() const;
   INTKEYBOARDINFO const *GetKeyboard() const;
+  /// TODO: 5822 public member process external event.
```
Comment (rc-swag) [inline core/src/kmx/kmx_processevent.h:109]:
```suggestion

```

File: core/src/kmx/kmx_processor.cpp [hunk 2]
```diff
@@ -103,6 +103,36 @@ kmx_processor::update_option(
...
+  switch (event){
+    case KM_KBP_EVENT_KEYBOARD_ACTIVATED:{
+      // reset any current actions in the queue as we have
+      // just loaded a new keyboard
+      _kmx.GetActions()->ResetQueue();
+      state->actions().clear();
+      // TODO: #5822
+      // just force it if caps always off is set.
+      // We need to pass the current modifiers in from the
+      // platform if we were to check.
```
Comment (rc-swag) [inline core/src/kmx/kmx_processor.cpp:122]:
```suggestion
```

File: core/src/kmx/kmx_processor.cpp [hunk 3]
```diff
@@ -103,6 +103,36 @@ kmx_processor::update_option(
...
+    KMX_DWORD dummy_modifiers = 0;
+    if (_kmx.GetKeyboard()->Keyboard->dwFlags & KF_CAPSALWAYSOFF) {
+       // TODO: #5822    need to use public member to access_kmx
```
Comment (rc-swag) [inline core/src/kmx/kmx_processor.cpp:125]:
```suggestion
```

File: core/src/kmx/kmx_processor.cpp [hunk 4]
```diff
@@ -103,6 +103,31 @@ kmx_processor::update_option(
   return option(scope, key, value);
 }
 
+km_kbp_status
+kmx_processor::external_event(
+      km_kbp_state* state,
+      uint32_t event,
+      void* _kmn_unused(data)
+      ) {
```
Comment (mcdurdin) [inline core/src/kmx/kmx_processor.cpp:111]:
```suggestion
  km_kbp_state* state,
  uint32_t event,
  void* _kmn_unused(data)
) {
```

File: core/src/km_kbp_processevent_api.cpp [hunk 5]
```diff
@@ -28,7 +28,7 @@ km_kbp_event(
   switch(event) {
     case KM_KBP_EVENT_KEYBOARD_ACTIVATED:
       assert(data == nullptr);
-      if(data == nullptr) {
+      if(data != nullptr) {
```
Comment (mcdurdin) [inline core/src/km_kbp_processevent_api.cpp:31]: What's happened here???

```suggestion
      if(data == nullptr) {
```

File: core/src/kmx/kmx_processevent.h [hunk 6]
```diff
@@ -107,6 +106,7 @@ class KMX_ProcessEvent {
   KMX_Environment *GetEnvironment();
   KMX_Environment const *GetEnvironment() const;
   INTKEYBOARDINFO const *GetKeyboard() const;
+  void SetCapsLock(KMX_DWORD &modifiers, KMX_BOOL capsLockOn, KMX_BOOL force = FALSE);
```
Comment (mcdurdin) [inline core/src/kmx/kmx_processevent.h:109]: Why did you move this down? It seems to belong better where it was previously on line 89.

File: core/src/kmx/kmx_processor.cpp [hunk 7]
```diff
@@ -103,6 +103,31 @@ kmx_processor::update_option(
...
+      // reset any current actions in the queue as a new keyboard
+      // has been activated
+      _kmx.GetActions()->ResetQueue();
+      state->actions().clear();
+    KMX_DWORD dummy_modifiers = 0;
+    if (_kmx.GetKeyboard()->Keyboard->dwFlags & KF_CAPSALWAYSOFF) {
+      _kmx.SetCapsLock(dummy_modifiers, FALSE, TRUE);
+      }
+    }
+    break;
```
Comment (mcdurdin) [inline core/src/kmx/kmx_processor.cpp:124]:
```suggestion
    case KM_KBP_EVENT_KEYBOARD_ACTIVATED:
      // reset any current actions in the queue as a new keyboard
      // has been activated
      _kmx.GetActions()->ResetQueue();
      state->actions().clear();
      if (_kmx.GetKeyboard()->Keyboard->dwFlags & KF_CAPSALWAYSOFF) {
        KMX_DWORD dummy_modifiers = 0;
        _kmx.SetCapsLock(dummy_modifiers, FALSE, TRUE);
      }
      break;
```

File: core/src/kmx/kmx_processor.cpp [hunk 8]
```diff
@@ -103,6 +103,31 @@ kmx_processor::update_option(
...
 }
 
+km_kbp_status
+kmx_processor::external_event(
+      km_kbp_state* state,
+      uint32_t event,
+      void* _kmn_unused(data)
+      ) {
+
+  switch (event){
```
Comment (mcdurdin) [inline core/src/kmx/kmx_processor.cpp:113]:
```suggestion
  switch (event) {
```

File: core/src/kmx/kmx_processor.hpp [hunk 9]
```diff
@@ -61,6 +61,13 @@ namespace kbp
       km_kbp_state *state
       ) override;
 
+    km_kbp_status
+    external_event(
+      km_kbp_state* state,
+      uint32_t event,
+      void* data
+      ) override;
```
Comment (mcdurdin) [inline core/src/kmx/kmx_processor.hpp:69]:
```suggestion
    ) override;

    km_kbp_status
    external_event(
      km_kbp_state* state,
      uint32_t event,
      void* data
    ) override;
```

File: core/tests/unit/kmx/kmx_external_event.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+#include <map>
+#include <iostream>
+#include <sstream>
+
+/** This test will test the infrastructure around the external event processing
+ *  The functions tested are:
+ *  km_kbp_event with the event KM_KBP_EVENT_KEYBOARD_ACTIVATED
+ *
+ */
```
Comment (mcdurdin) [inline core/tests/unit/kmx/kmx_external_event.cpp:26]:
```suggestion
/** 
 * This test will test the infrastructure around the external event processing
 * The functions tested are:
 * - km_kbp_event with the event KM_KBP_EVENT_KEYBOARD_ACTIVATED
 */
```

File: core/tests/unit/kmx/kmx_external_event.cpp [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+using namespace km::kbp::kmx;
+
+km_kbp_option_item test_env_opts[] =
+{
+  KM_KBP_OPTIONS_END
+};
+
+int error_args() {
+    std::cerr << "kmx: Not enough arguments." << std::endl;
+    return 1;
```
Comment (mcdurdin) [inline core/tests/unit/kmx/kmx_external_event.cpp:37]:
```suggestion
  std::cerr << "kmx: Not enough arguments." << std::endl;
  return 1;
```

File: windows/src/engine/keyman32/kmprocessactions.cpp [hunk 12]
```diff
@@ -198,7 +199,33 @@ ProcessActionsNonUpdatableParse(BOOL* emitKeyStroke) {
...
+  return TRUE;
+}
+
+BOOL
+ProcessActionsExternalEvent() {
+  PKEYMAN64THREADDATA _td = ThreadGlobals();
+  if (!_td) {
+    return FALSE;
+  }
+  // Currently the only Action handled is KM_KBP_IT_CAPSLOCK.
```
Comment (mcdurdin) [inline windows/src/engine/keyman32/kmprocessactions.cpp:221]:
```suggestion
  // Currently only a subset of actions are handled.
```

File: windows/src/engine/keyman32/kmprocessactions.h [hunk 13]
```diff
@@ -26,4 +26,15 @@ BOOL ProcessActions(BOOL* emitKeyStroke);
  */
 BOOL ProcessActionsNonUpdatableParse(BOOL* emitKeyStroke);
 
+/**
+ * This function processes the actions queued in the core processor in
+ * the external event case. That is actions not caused by a keystroke but from selecting a
+ * different keyboard for example.
+ * Currently only CAPS Lock and invalidate key stroke are processed.
```
Comment (mcdurdin) [inline windows/src/engine/keyman32/kmprocessactions.h:33]:
```suggestion
 * Currently only KM_KBP_IT_CAPSLOCK and KM_KBP_IT_INVALIDATE_CONTEXT are processed.
```

File: windows/src/engine/keyman32/kmprocessactions.h [hunk 14]
```diff
@@ -26,4 +26,15 @@ BOOL ProcessActions(BOOL* emitKeyStroke);
...
+ *
+ * @return BOOL  True if actions were successfully processed
```
Comment (mcdurdin) [inline windows/src/engine/keyman32/kmprocessactions.h:35]:
```suggestion
 * @return BOOL  TRUE if actions were successfully processed
```

**Review by mcdurdin — APPROVED** [review]
LGTM, only nits :grin:

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_CAPSOFF-EXT_EVENT (PASSED):** Installed the attached PR build (Keyman 17.0.143-alpha-test-9353) in Windows 10 OS (Native Machine) and here is my observation: 1. Installed caps_always_off keyboard. 2. Verified thta the caps lock indicator turned off in the keyboard. 
* **TEST_CAPSOFF-1 (PASSED):** Verified that after releasing a key, it shows ncaps_little_a in the Notepad.
* **TEST_CAPSOFF-2 (PASSED):** Verified that the caps lock indicator stays turned off and the outputs results ncaps_little_a.
* **TEST_CAPSOFF-3 (PASSED):** Verified that the output shows ncaps_little_a in the Notepad.
* **TEST_CAPSOFF-4 (PASSED):** Verified that the output results ncaps_shift_A in the Notepad.

   


* **TEST_CAPSONLY-1 (PASSED):** 1. Installed ShiftFreesCaps keyboard. 2. Verified that the result shows pass in the Notepad after releasing 1 number key. 
* **TEST_CAPSONLY-2 (PASSED):** 1. Verified that the caps lock indicator turned on. 2. Output shows Pass.
* **TEST_CAPSONLY-3 (PASSED):** 1. Verified that the caps lock indicator turned on. 2. Output shows Pass.
* **TEST_CAPSONLY-4 (PASSED):** 1. Verified that the caps lock indicator turned off. 2. Output shows Pass.
* **TEST_CAPSONLY-5 (PASSED):** 1. Verified that the caps lock indicator turned off. 2. No output appear in the Notepad app.

File: core/src/km_kbp_processevent_api.cpp (hunk 5 above)
Comment (rc-swag) [reply core/src/km_kbp_processevent_api.cpp:31]: It was incorrect. Look at line 30. Line 31 will only be reached if NDEBUG is defined.

File: core/src/kmx/kmx_processevent.h (hunk 6 above)
Comment (rc-swag) [reply core/src/kmx/kmx_processevent.h:109]: I exposed the method it is now public. I could make a new public member that just calls the private SetCapsLock Method.

File: core/src/km_kbp_processevent_api.cpp (hunk 5 above)
Comment (mcdurdin) [reply core/src/km_kbp_processevent_api.cpp:31]: Yeah, easy to do, assert(true_condition) followed by if(false_condition) ... almost want a pattern of:

```
assert(!(data != nullptr));
if(data != nullptr) {
  ...
}
```

File: core/src/kmx/kmx_processevent.h (hunk 6 above)
Comment (mcdurdin) [reply core/src/kmx/kmx_processevent.h:109]: I see

Comment (mcdurdin) [reply core/src/kmx/kmx_processevent.h:109]: No, it's fine to keep SetCapsLock public if that's what it needs to be.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.158-alpha](https://keyman.com/downloads/releases/alpha/17.0.158)
